x-pack/plugins/security_solution/docs/prebuilt_rules_customization_rfc.md

```diff
@@ -0,0 +1,2867 @@
+      <td>
+        <pre> 
+          {
+            type: 'external',
+            isCustomized: false,
+              ...
+          }
+          </pre>
+      </td>
+      <td><pre>false</pre></td>
+      <td>Yes</td>
+      <td>
+        <pre>
+          {
+            rule_source: {
+              type: 'external',
+              isCustomized: true,
+              ...
+            },
+            immutable: true,
+          }
+        </pre>
+      </td>
+    </tr>
+
+  </tbody>
+</table>
+
+As explained before, all these endpoints suffer from the tightly-coupled logic problem described in the section [`Problem with tightly coupled logic in our endpoints`](#problem-with-tightly-coupled-logic-in-our-endpoints).
+
+We should therefore create new CRUD methods as needed to uncouple them logically and cleanly apply migrations to them.
+
+---
+
+#### Bulk editing rules
+
+The endpoint **Bulk Actions** - `POST /rules/_bulk_action` has the same outputs from the migration logic as the endpoints listed in the above section, so it was included in the scenarios in the table above. 
+
+However, this endpoint has some specific details that should be mentioned.
+
+Firstly, when bulk editing rules, the migration should be carried out for the following use cases:
+  - Bulk adding or deleting index patterns
+  - Bulk adding or deleting tags
+  - Updating rule schedules
+  - Adding rules actions
+  - Duplicating a rule (with the migration applied only to the new rule)
+
+Out of the actions mentioned above, the only use cases that should possibly result in the migration having a result of `rule_source.isCustomized = true` are the first three:
+  - Bulk adding or deleting index patterns
+  - Bulk adding or deleting tags
+  - Updating rule schedules
+
+That means that updating a rule's actions or duplicating a rule should not be considered a customization of a prebuilt rule.
+
+Secondly, in contrast with other endpoints, the migration for bulk editing rules needs to be carried within the Alerting Framework's `RuleClient` since we need to migrate the rule's `rule_source` and `immutable` parameters before the `RuleClient` does the saving of the rule into Elasticsearch.
+
+This however, gives us the advantage that the `RuleClient` already includes logic to calculate if a rule's attributes or parameters have been modified from its initial values, so we can rely on that to calculate the `rule_source.is_customized` field during the update.
+
+The `RulesClient` class has a `bulkEdit` method, which is called by our **Bulk Actions endpoint** `POST /rules/_bulk_action`. That method includes complex logic, but we can focus on the `updateRuleAttributesAndParamsInMemory` method, where the rule attributes and parameters are updated before being saved to ES. This function also calculates the booleans `isAttributesUpdateSkipped` and `isParamsUpdateSkipped` which we can leverage to calculate the new value of the `rule_source.is_customized` field in our params.
```

Comment:
Yes, that's true. But here I'm referring to the calculation of the `is_customized` field. The update of fields does happen within the Alerting Framework, and if we want to know if a value was indeed updated so that it has diverged from its base version, we need to that check within the logic of the Rules Client.

Reply:
After looking into the code, I think we should do it here, somewhere inside the `paramsModifier` callback:

https://github.com/elastic/kibana/blob/a7c8ef421ee097810af51413b621871874becfa4/x-pack/plugins/security_solution/server/lib/detection_engine/rule_management/logic/bulk_actions/bulk_edit_rules.ts#L30-L62

But not inside the `updateRuleAttributesAndParamsInMemory` on the framework side.

Reply:
The problem is that for calculating `is_customized` we need to know what the current values of the rule SO are. In the Security Solution side (i.e.: `x-pack/plugins/security_solution/server/lib/detection_engine/rule_management/logic/bulk_actions/bulk_edit_rules.ts`), we just create the set of operations that modify the attributes and params of the rules and pass them over to the Alerting Framework's `bulkEdit` method.

Only in the Alerting Framework is the rule SO fetched and its params and atributes modified. 
So basically, the Security Solution side nevers sees the current values of the rule's fields. It only recieves the already modified values, in `result`.
This means that if we want to do the comparison of the current and resulting values of the rule's fields, we have two options:
1. Do the comparison in the Alerting Framework side, where the rule is fetched and the modifications applied.
2. We need to fetch the rule SO on the Security Solution side, -which currently does not happen-, and then compare the current values with the resulting values. 

Now I think that this second options is preferable, to avoid adding logic to the Alerting Framework side, but we might need to consider the performance overhead of fetching the rule SO.

Reply:
Calculation of `is_customized` options:
1. Modify the rulesParamsModifier to take as parameter the whole rule, or additionally the operations.
2. Do it as part of the `validateMutatedRuleTypeParams` method `x-pack/plugins/alerting/server/lib/validate_mutated_rule_type_params.ts`